src/lib.rs

```diff
@@ -380,6 +515,142 @@ impl Invariant {
         }
     }
 
+    fn calculate_swap(
+        &self,
+        pool_key: PoolKey,
+        x_to_y: bool,
+        amount: TokenAmount,
+        by_amount_in: bool,
+        sqrt_price_limit: SqrtPrice,
+    ) -> Result<CalculateSwapResult, InvariantError> {
+        let current_timestamp = exec::block_timestamp();
+
+        if amount.is_zero() {
+            return Err(InvariantError::AmountIsZero);
+        }
+
+        let mut ticks: Vec<Tick> = vec![];
+
+        let mut pool = self.pools.get(&pool_key)?;
+
+        if x_to_y {
+            if pool.sqrt_price <= sqrt_price_limit
+                || sqrt_price_limit > SqrtPrice::new(MAX_SQRT_PRICE)
+            {
+                return Err(InvariantError::WrongLimit);
+            }
+        } else if pool.sqrt_price >= sqrt_price_limit
+            || sqrt_price_limit < SqrtPrice::new(MIN_SQRT_PRICE)
+        {
+            return Err(InvariantError::WrongLimit);
+        }
+
+        let mut remaining_amount = amount;
+
+        let mut total_amount_in = TokenAmount(0);
+        let mut total_amount_out = TokenAmount(0);
+
+        let event_start_sqrt_price = pool.sqrt_price;
+        let mut event_fee_amount = TokenAmount(0);
+
+        while !remaining_amount.is_zero() {
+            let (swap_limit, limiting_tick) = self.tickmap.get_closer_limit(
+                sqrt_price_limit,
+                x_to_y,
+                pool.current_tick_index,
+                pool_key.fee_tier.tick_spacing,
+                pool_key,
+            )?;
+            let result = unwrap!(compute_swap_step(
+                pool.sqrt_price,
+                swap_limit,
+                pool.liquidity,
+                remaining_amount,
+                by_amount_in,
+                pool_key.fee_tier.fee,
+            ));
+
+            let is_enough_amount_to_cross = unwrap!(is_enough_amount_to_change_price(
+                remaining_amount,
+                result.next_sqrt_price,
+                pool.liquidity,
+                pool_key.fee_tier.fee,
+                by_amount_in,
+                x_to_y,
+            ));
+
+            // make remaining amount smaller
+            if by_amount_in {
+                remaining_amount -= result.amount_in + result.fee_amount;
+            } else {
+                remaining_amount -= result.amount_out;
+            }
+
+            unwrap!(pool.add_fee(result.fee_amount, x_to_y, self.config.protocol_fee));
+            event_fee_amount += result.fee_amount;
+
+            pool.sqrt_price = result.next_sqrt_price;
+
+            total_amount_in += result.amount_in + result.fee_amount;
+            total_amount_out += result.amount_out;
+
+            // Fail if price would go over swap limit
+            if pool.sqrt_price == sqrt_price_limit && !remaining_amount.is_zero() {
+                return Err(InvariantError::PriceLimitReached);
+            }
+
+            if let Some((tick_index, is_initialized)) = limiting_tick {
```

Comment:
I think you should refactor entire `if` here. `pool.cross_tick` now handle all cases:
1. crossing initialized tick
2. crossing empty tick (tick resulting from a limited scope of searching)
3. calculating tick based on `sqrt_price`(if you're not moving through initialized tick and empty tick)

Also, note that you are calculating `is_enough_amount_to_change_price` twice